File: bindings/swift/Sources/DDS/DDS.swift
// DDS Swift bindings — C interop wrapper for libdds_ffi.
//
// Usage:
//   let ident = try DDS.Identity.create(label: "alice")
//   print(ident.urn)
//
// Link against libdds_ffi.dylib / .so.

import Foundation

// Import the C header module
@_implementationOnly import CDDS

/// Error codes from the DDS library.
public enum DDSErrorCode: Int32 {
    case ok = 0
    case invalidInput = -1
    case crypto = -2
    case token = -3
    case trust = -4
    case policyDenied = -5
    case `internal` = -99
}

/// Error thrown by DDS operations.
public struct DDSError: Error, CustomStringConvertible {
    public let code: DDSErrorCode
    public let detail: String
    public var description: String { "DDS error \(code): \(detail)" }
}

// MARK: - Internal Helpers

private func callJSON(_ block: (UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>) -> Int32) throws -> [String: Any] {
    var out: UnsafeMutablePointer<CChar>? = nil
    let rc = block(&out)
    var json: [String: Any] = [:]
    if let ptr = out {
        let str = String(cString: ptr)
        dds_free_string(ptr)
        if let data = str.data(using: .utf8),
           let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            json = parsed
        }
    }
    guard rc == DDSErrorCode.ok.rawValue else {
        throw DDSError(code: DDSErrorCode(rawValue: rc) ?? .internal,
                       detail: json["error"] as? String ?? "")
    }
    return json
}

// MARK: - Public API

public enum DDS {

    /// Get the library version.
    public static func version() throws -> String {
        var out: UnsafeMutablePointer<CChar>? = nil
        let rc = dds_version(&out)
        guard rc == DDSErrorCode.ok.rawValue, let ptr = out else {
            throw DDSError(code: .internal, detail: "version failed")
        }
        let v = String(cString: ptr)
        dds_free_string(ptr)
        return v
    }

    // MARK: Identity
    public enum Identity {
        public static func create(label: String) throws -> [String: Any] {
            try callJSON { dds_identity_create(label, $0) }
        }

        public static func createHybrid(label: String) throws -> [String: Any] {
            try callJSON { dds_identity_create_hybrid(label, $0) }
        }

        public static func parseUrn(_ urn: String) throws -> [String: Any] {
            try callJSON { dds_identity_parse_urn(urn, $0) }
        }
    }

    // MARK: Token
    public enum Token {
        public static func createAttest(config: [String: Any]) throws -> [String: Any] {
            let data = try JSONSerialization.data(withJSONObject: config)
            let json = String(data: data, encoding: .utf8)!
            return try callJSON { dds_token_create_attest(json, $0) }
        }

        public static func validate(tokenCborHex: String) throws -> [String: Any] {
            try callJSON { dds_token_validate(tokenCborHex, $0) }
        }
    }

    // MARK: Policy
    public enum Policy {
        public static func evaluate(config: [String: Any]) throws -> [String: Any] {
            let data = try JSONSerialization.data(withJSONObject: config)
            let json = String(data: data, encoding: .utf8)!
            return try callJSON { dds_policy_evaluate(json, $0) }
        }
    }
}

File: bindings/kotlin/src/main/kotlin/vouchsafe/dds/DDS.kt
// DDS Kotlin bindings — JNA wrapper for libdds_ffi.
//
// Usage:
//   val client = DDS()
//   val ident = client.identityCreate("alice")
//   println(ident.getString("urn"))
//
// Requires JNA and org.json on classpath.
// Native library must be in java.library.path.

package vouchsafe.dds

import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.ptr.PointerByReference
import org.json.JSONObject

/** Error codes from the DDS library. */
object DDSError {
    const val OK = 0
    const val INVALID_INPUT = -1
    const val CRYPTO = -2
    const val TOKEN = -3
    const val TRUST = -4
    const val POLICY_DENIED = -5
    const val INTERNAL = -99
}

class DDSException(val code: Int, message: String) : Exception("DDS error $code: $message")

/** JNA interface for the native library. */
internal interface DDSNative : Library {
    fun dds_identity_create(label: String, out: PointerByReference): Int
    fun dds_identity_create_hybrid(label: String, out: PointerByReference): Int
    fun dds_identity_parse_urn(urn: String, out: PointerByReference): Int
    fun dds_token_create_attest(configJson: String, out: PointerByReference): Int
    fun dds_token_validate(tokenHex: String, out: PointerByReference): Int
    fun dds_policy_evaluate(configJson: String, out: PointerByReference): Int
    fun dds_version(out: PointerByReference): Int
    fun dds_free_string(s: Pointer)
}

/** DDS client — wraps the native library. */
class DDS(libName: String = "dds_ffi") {
    private val lib: DDSNative = Native.load(libName, DDSNative::class.java)

    private fun callJson(block: (PointerByReference) -> Int): JSONObject {
        val out = PointerByReference()
        val rc = block(out)
        val ptr = out.value
        val json = if (ptr != null) ptr.getString(0) else "{}"
        if (ptr != null) lib.dds_free_string(ptr)
        if (rc != DDSError.OK) throw DDSException(rc, json)
        return JSONObject(json)
    }

    fun version(): String {
        val out = PointerByReference()
        val rc = lib.dds_version(out)
        if (rc != DDSError.OK) throw DDSException(rc, "version failed")
        val v = out.value.getString(0)
        lib.dds_free_string(out.value)
        return v
    }

    fun identityCreate(label: String): JSONObject =
        callJson { lib.dds_identity_create(label, it) }

    fun identityCreateHybrid(label: String): JSONObject =
        callJson { lib.dds_identity_create_hybrid(label, it) }

    fun identityParseUrn(urn: String): JSONObject =
        callJson { lib.dds_identity_parse_urn(urn, it) }

    fun tokenCreateAttest(config: JSONObject): JSONObject =
        callJson { lib.dds_token_create_attest(config.toString(), it) }

    fun tokenValidate(tokenCborHex: String): JSONObject =
        callJson { lib.dds_token_validate(tokenCborHex, it) }

    fun policyEvaluate(config: JSONObject): JSONObject =
        callJson { lib.dds_policy_evaluate(config.toString(), it) }
}
